### DISCUSSION TITLE
Is there a way to wrap the code block inside monitor block?

### DISCUSSION THREAD

**Discussion by rishabkshk** [codefori/vscode-ibmi#1345] [General]
Can I select a code block and wrap it inside a monitor block and then the cursor directly moves to on-error for me to add the handling of the exception? I tried to use snippets but it replaces the actual selected code with monitor block. This works for commenting multiple lines. Something similar like below java code for example.

![image](https://github.com/halcyon-tech/vscode-ibmi/assets/21074746/c1657537-2409-484c-b957-ee8390d894e4)

**Comment by sebjulliand** [comment]
@worksofliam I'll try something...

**Reply by worksofliam** [reply]
I believe there is a way to do this with snippets!

**Reply by sebjulliand** [reply]
Too late, I had to try 😅
![surround](https://github.com/halcyon-tech/vscode-ibmi/assets/11096890/c1f0b2c9-9b67-4ccf-aa06-faa7c4ab94a0)

**Reply by chrjorgensen** [reply]
@sebjulliand But where is the PR - or extension link?! 😉

**Reply by sebjulliand** [reply]
@chrjorgensen it's on its way my Lord 😁

**Reply by rishabkshk** [reply]
> @chrjorgensen it's on its way my Lord 😁

Is it customizable with a custom snippet? Thanks for making this available.

**Reply by sebjulliand** [reply]
Not now but maybe later!

**Reply by rishabkshk** [reply]
> Not now but maybe later!

Thanks again! Btw for custom snippets, I got the extension in [this ](https://github.com/orgs/halcyon-tech/discussions/1345#discussioncomment-6112210)discussion working.

**Reply by sebjulliand** [reply]
That's good to know!
I decided to have a custom implementation for your need because it allows to check if the code you want to wrap is Free RPG, and it will also surround a whole selection or a line, depending on what you have selected in the editor when running the action - I wanted it to be clever with RPGLE 😉

**Reply by rishabkshk** [reply]
> That's good to know! I decided to have a custom implementation for your need because it allows to check if the code you want to wrap is Free RPG, and it will also surround a whole selection or a line, depending on what you have selected in the editor when running the action - I wanted it to be clever with RPGLE 😉

Which means it will be surrounded by fixed/free as per the code block? That's visionary.👏

**Reply by sebjulliand** [reply]
It can only surround Free RPGLE; but it will try to do it in a clever way 😊
Here goes: https://github.com/halcyon-tech/vscode-rpgle/pull/223

**Comment by Wright4i** [comment]
I'd recommend looking at the extension marketplace. "Surround With" seems
like it'd do the trick.

**Reply by chrjorgensen** [reply]
Did a quick search in the extension marketplace for "surround with", and this extension looks the most promising with configurable snippets:
https://marketplace.visualstudio.com/items?itemName=yatki.vscode-surround

![image](https://github.com/halcyon-tech/vscode-ibmi/assets/13275072/8ebe9d04-ab49-4dd6-9268-65ad79dec04e)

**Reply by chrjorgensen** [reply]
Digging into this, it seems like snippets can do it (as @worksofliam wrote) by using the variable `TM_SELECTED_TEXT` in the snippet.
See more here: https://code.visualstudio.com/docs/editor/userdefinedsnippets#_variables

This snippet will do the trick:

```json
  "monitor": {
    "prefix": "monitor",
    "body": [
      "monitor;\n  $TM_SELECTED_TEXT\non-error $0;\nendmon;"
    ],
    "description": "Surround block with MONITOR"
  }
```

Select your statement, press F1 and type `snip` and select the command `Snippets: Insert Snippet` and choose the one called `monitor`. Your statement will now be surrounded by `monitor/on-error/endmon` and the cursor is placed at the `on-error` statement.

So installing an extension is not required to get this functionality.

**Reply by rishabkshk** [reply]
I was able to define a custom snippet using the extension you described. Below is my snippet.

```
"surround.custom": {
        "Monitor": {
            "label": "monitor",
            "description": "Wrap inside monitor-block",
            "snippet": "Monitor; \n $TM_SELECTED_TEXT \n On-Error;  \n   Exsr ErrorLog;\nEndMon;\n", 
            "languageIds": [ "*" ]
        }
    },
```

I will check for vscode's inbuilt snippets method too.
